tutorials/intermediate/blitting.py

```diff
@@ -0,0 +1,186 @@
+# The previous example works and shows a simple animation, however
+# because we are only grabbing the background once, if the size or dpi
+# of the figure change, the background will be invalid and result in
+# incorrect images.  There is also a global variable and a fair amount
+# of boiler plate which suggests we should wrap this in a class.
+#
+# Class-based example
+# -------------------
+#
+# We can use a class to encapsulate the boilerplate logic and state of
+# restoring the background, drawing the artists, and then blitting the
+# result to the screen.  Additionally, we can use the ``'draw_event'``
+# callback to capture a new background whenever a full re-draw
+# happens.
+
+
+class BlitManager:
+    def __init__(self, canvas, animated_artists=()):
+        """
+        Parameters
+        ----------
+        canvas : CanvasAgg
+            The canvas to work with, this only works for sub-classes of the Agg
+            canvas which have the `~CanvasAgg.copy_from_bbox` and
+            `~CanvasAgg.restore_region` methods.
+
+        animated_artists : Iterable[Artist]
+            List of the artists to manage
+        """
+        self.canvas = canvas
+        self._bg = None
+        self._artists = []
+
+        for a in animated_artists:
+            self.add_artist(a)
+        # grab the background on every draw
+        self.cid = canvas.mpl_connect("draw_event", self.on_draw)
+
+    def on_draw(self, event):
+        """Callback to register with 'draw_event'."""
+        cv = self.canvas
+        if event is not None:
+            if event.canvas != cv:
+                raise RuntimeError
+        self._bg = cv.copy_from_bbox(cv.figure.bbox)
+        self._draw_animated()
+
+    def add_artist(self, art):
+        """
+        Add an artist to be managed.
+
+        Parameters
+        ----------
+        art : Artist
+
+            The artist to be added.  Will be set to 'animated' (just
+            to be safe).  *art* must be in the figure associated with
+            the canvas this class is managing.
+
+        """
+        if art.figure != self.canvas.figure:
+            raise RuntimeError
+        art.set_animated(True)
+        self._artists.append(art)
+
+    def _draw_animated(self):
+        """Draw all of the animated artists."""
+        fig = self.canvas.figure
+        for a in self._artists:
+            fig.draw_artist(a)
+
+    def update(self):
+        """Update the screen with animated artists."""
+        cv = self.canvas
+        fig = cv.figure
+        # paranoia in case we missed the draw event,
+        if self._bg is None:
+            self.on_draw(None)
+        else:
+            # restore the old background
+            cv.restore_region(self._bg)
+            # draw all of the animated artists
+            self._draw_animated()
+            # update the screen
+            cv.blit(fig.bbox)
+        # let the GUI event loop process anything it has to do
+        cv.flush_events()
```

Comment:
I've never worked with the event loop. Is there some incantation like

```python
fig.draw()
cv.flush_events()
```

that could be used instead of `plt.pause` in the original simple example? It would make a lot more semantic sense (and avoid confusion of introducing new concepts in the class version that weren't used in the simpler example above).

Reply:
No, I don't think so.  What `plt.pause` is primarily doing above is managing the event loop (because in the first example we know that the script is in charge so it is ok to start / stop the event loop).  In this case it is meant to be a more general-purpose bit of code that can easily be embedded in other contexts so we should do the least invasive thing (which is to spin the event loop "by hand").

Reply:
Below you write "Probably should replace this with lower-level things to get the figure on the screen."

That's probably all I was looking for, actually.

Reply:
Ended up going "higher level", but I think it is clearer now.